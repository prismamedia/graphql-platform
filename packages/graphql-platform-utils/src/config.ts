export * from './config/name.js';
export * from './config/optional-deprecation.js';
export * from './config/optional-description.js';
export * from './config/optional-flag.js';
