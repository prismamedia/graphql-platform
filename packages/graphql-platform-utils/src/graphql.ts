import assert from 'assert';
import {
  ArgumentNode,
  GraphQLArgumentConfig,
  GraphQLEnumType,
  GraphQLLeafType,
  GraphQLList,
  GraphQLNonNull,
  GraphQLResolveInfo,
  GraphQLScalarType,
  GraphQLType,
  isLeafType,
  isListType,
  isNonNullType,
  isOutputType,
  isScalarType,
  isSchema,
  isWrappingType,
  ValueNode,
} from 'graphql';
import { pathToArray } from 'graphql/jsutils/Path';
import { isEqual } from 'lodash';
import { isIterable } from '.';
import { UnexpectedValueError } from './errors';
import { isPlainObject, PlainObject } from './object';
import { addPath, Path } from './path';

export type GraphQLArgumentConfigMap<TArgs extends PlainObject | undefined> = {
  readonly [argName in keyof TArgs]: GraphQLArgumentConfig;
};

export const GraphQLNonNullDecorator = <T extends GraphQLType>(
  type: T,
  nonNull: boolean,
) =>
  isNonNullType(type)
    ? nonNull
      ? type
      : type.ofType
    : nonNull
    ? GraphQLNonNull(type)
    : type;

export const GraphQLListDecorator = <T extends GraphQLType>(
  type: T,
  list: boolean,
) => (list ? GraphQLList(type) : type);

export function parseValueNode(
  value: ValueNode,
  variables?: GraphQLResolveInfo['variableValues'],
): any {
  switch (value.kind) {
    case 'BooleanValue':
      return value.value;

    case 'IntValue':
      return Number.parseInt(value.value, 10);

    case 'FloatValue':
      return Number.parseFloat(value.value);

    case 'Variable':
      assert(isPlainObject(variables));
      assert(value.name.value in variables);

      return variables[value.name.value];

    case 'ObjectValue':
      return Object.fromEntries(
        value.fields.map((field) => [
          field.name.value,
          parseValueNode(field.value, variables),
        ]),
      );

    case 'ListValue':
      return value.values.map((value) => parseValueNode(value, variables));

    case 'NullValue':
      return null;

    default:
      return value.value;
  }
}

export const parseArgumentNodes = (
  args: ReadonlyArray<ArgumentNode>,
  variables?: GraphQLResolveInfo['variableValues'],
): PlainObject =>
  Object.fromEntries(
    args.map(({ name, value }) => [
      name.value,
      parseValueNode(value, variables),
    ]),
  );

export function getResolverPath(info: GraphQLResolveInfo): Path {
  return pathToArray(info.path).reduce(addPath, undefined) as Path;
}

export function isGraphQLResolveInfo(
  value: unknown,
): value is GraphQLResolveInfo {
  return (
    isPlainObject(value) &&
    'fieldName' in value &&
    typeof value.fieldName === 'string' &&
    'fieldNodes' in value &&
    Array.isArray(value.fieldNodes) &&
    'returnType' in value &&
    isOutputType(value.returnType) &&
    'parentType' in value &&
    isOutputType(value.parentType) &&
    'path' in value &&
    'schema' in value &&
    isSchema(value.schema)
  );
}

export function assertScalarValue(
  type: GraphQLScalarType,
  value: any,
  path?: Path,
) {
  let parsedValue;

  try {
    parsedValue = type.parseValue(value);
  } catch (error) {
    throw new UnexpectedValueError(value, `a "${type.name}"`, path);
  }

  // Should never happen
  if (parsedValue == null) {
    throw new UnexpectedValueError(parseValueNode, `a "${type.name}"`, path);
  }

  return parsedValue;
}

export function assertEnumValue(
  type: GraphQLEnumType,
  value: any,
  path?: Path,
) {
  const enumValue = type
    .getValues()
    .find((enumValue) => isEqual(enumValue.value, value));

  if (enumValue === undefined) {
    throw new UnexpectedValueError(
      value,
      `a value among "${type
        .getValues()
        .map(({ value }) => value)
        .join(', ')}"`,
      path,
    );
  }

  return enumValue.value;
}

export function assertLeafValue(
  type: GraphQLLeafType,
  value: any,
  path?: Path,
) {
  return isScalarType(type)
    ? assertScalarValue(type, value, path)
    : assertEnumValue(type, value, path);
}

export function isWrappedLeafType(
  maybeWrappedLeafType: unknown,
): maybeWrappedLeafType is GraphQLType {
  return isWrappingType(maybeWrappedLeafType)
    ? isWrappedLeafType(maybeWrappedLeafType.ofType)
    : isLeafType(maybeWrappedLeafType);
}

export function assertWrappedLeafValue(
  type: GraphQLType,
  value: any,
  path?: Path,
): any {
  if (value === undefined) {
    throw new UnexpectedValueError(value, `a "${type}"`, path);
  } else if (value === null) {
    if (isNonNullType(type)) {
      throw new UnexpectedValueError(value, `a "${type}"`, path);
    }

    return value;
  }

  if (isNonNullType(type)) {
    return assertWrappedLeafValue(type.ofType, value, path);
  } else if (isListType(type)) {
    if (!isIterable(value)) {
      throw new UnexpectedValueError(value, `a "${type}"`, path);
    }

    return Array.from(value, (value, index) =>
      assertWrappedLeafValue(type.ofType, value, addPath(path, index)),
    );
  } else if (isLeafType(type)) {
    return assertLeafValue(type, value, path);
  }
}
