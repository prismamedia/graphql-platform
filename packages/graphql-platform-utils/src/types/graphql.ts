export {
  FieldConfig as GraphQLFieldConfig,
  FieldConfigMap as GraphQLFieldConfigMap,
  FieldResolver as GraphQLFieldResolver,
  FieldResolverArgs as GraphQLFieldResolverArgs,
} from './graphql/field';
export { OperationType as GraphQLOperationType } from './graphql/operation-type';
