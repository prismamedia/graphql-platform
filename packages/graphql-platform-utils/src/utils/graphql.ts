export * from './graphql/get-enum-type';
export * from './graphql/list-decorator';
export * from './graphql/non-null-decorator';
export * from './graphql/parse-resolver-info';
