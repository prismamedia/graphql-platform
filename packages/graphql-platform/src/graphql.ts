import type * as graphql from 'graphql';
import type { ConnectorInterface } from './connector-interface.js';
import type { GraphQLPlatform } from './index.js';

export type GPBoundGraphQLFieldResolver<
  TRequestContext extends object,
  TConnector extends ConnectorInterface,
  TSource,
  TArgs = any,
  TResult = unknown,
> = (
  this: GraphQLPlatform<TRequestContext, TConnector>,
  source: TSource,
  args: TArgs,
  context: TRequestContext,
  info: graphql.GraphQLResolveInfo,
) => TResult;

export interface GPBoundGraphQLFieldConfig<
  TRequestContext extends object,
  TConnector extends ConnectorInterface,
  TSource,
  TArgs = any,
  TResult = unknown,
> extends graphql.GraphQLFieldConfig<TSource, TRequestContext, TArgs> {
  resolve?: GPBoundGraphQLFieldResolver<
    TRequestContext,
    TConnector,
    TSource,
    TArgs,
    TResult
  >;
}
