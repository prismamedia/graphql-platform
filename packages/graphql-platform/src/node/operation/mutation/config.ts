import type * as utils from '@prismamedia/graphql-platform-utils';
import type { ConnectorInterface } from '../../../connector-interface.js';
import type { CreationConfig } from './abstract-creation.js';
import type { DeletionConfig } from './abstract-deletion.js';
import type { UpdateConfig } from './abstract-update.js';

export type {
  CreationConfig,
  PostCreateArgs,
  PreCreateArgs,
} from './abstract-creation.js';
export type {
  DeletionConfig,
  PostDeleteArgs,
  PreDeleteArgs,
} from './abstract-deletion.js';
export type {
  PostUpdateArgs,
  PreUpdateArgs,
  UpdateConfig,
} from './abstract-update.js';

export interface MutationConfig<
  TRequestContext extends object,
  TConnector extends ConnectorInterface,
  TContainer extends object,
> {
  [utils.MutationType.CREATION]: CreationConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
  [utils.MutationType.UPDATE]: UpdateConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
  [utils.MutationType.DELETION]: DeletionConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
}
