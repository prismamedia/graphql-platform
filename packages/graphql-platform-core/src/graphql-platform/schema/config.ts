import { GraphQLOperationType } from '@prismamedia/graphql-platform-utils';
import { GraphQLFieldConfigMap, GraphQLObjectType, GraphQLSchemaConfig as BaseGraphQLSchemaConfig } from 'graphql';
import { OperationType } from '../operation';
import { ResourceMap } from '../resource/map';

export interface GraphQLSchemaConfigConfig
  extends Omit<BaseGraphQLSchemaConfig, 'mutation' | 'query' | 'subscription'> {
  resourceMap: ResourceMap;
  queryMap: GraphQLFieldConfigMap<any, any>;
  mutationMap: GraphQLFieldConfigMap<any, any>;
}

export class GraphQLSchemaConfig implements BaseGraphQLSchemaConfig {
  public constructor(readonly config: GraphQLSchemaConfigConfig) {}

  public getOperationTypeFieldMap<TType extends OperationType>(type: TType): GraphQLFieldConfigMap<any, any> {
    const fields: GraphQLFieldConfigMap<any, any> = {};

    for (const resource of this.config.resourceMap.values()) {
      for (const operation of resource.getOperations(type)) {
        if (operation.isPublic()) {
          fields[operation.name] = operation.getGraphQLFieldConfig();
        }
      }
    }

    return fields;
  }

  public get mutation() {
    return new GraphQLObjectType({
      name: 'Mutation',
      fields: { ...this.getOperationTypeFieldMap(GraphQLOperationType.Mutation), ...this.config.mutationMap },
    });
  }

  public get query() {
    return new GraphQLObjectType({
      name: 'Query',
      fields: { ...this.getOperationTypeFieldMap(GraphQLOperationType.Query), ...this.config.queryMap },
    });
  }

  public get subscription() {
    return undefined;
  }
}
