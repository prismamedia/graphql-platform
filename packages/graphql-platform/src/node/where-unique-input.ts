import {
  addPath,
  assertLeafValue,
  GraphQLNonNullDecorator,
  isPlainObject,
  Path,
  UnexpectedNullValueError,
  UnexpectedUndefinedValueError,
  UnexpectedValueError,
} from '@prismamedia/graphql-platform-utils';
import { Memoize } from '@prismamedia/ts-memoize';
import assert from 'assert';
import { GraphQLInputObjectType } from 'graphql';
import { ValueOf } from 'type-fest';
import { Node } from '../node';
import { Leaf, TComponent, TLeafValue } from './component';
import { UniqueConstraint } from './unique-constraint';

export type TWhereUniqueInputValue = {
  [componentName: string]: null | TLeafValue | TWhereUniqueInputValue;
};

export type TParsedWhereUniqueInputValue = TWhereUniqueInputValue;

export class WhereUniqueInput {
  public readonly public: boolean;
  public readonly name: string;

  public constructor(public readonly node: Node) {
    this.public = node.public;
    this.name = `${node.name}WhereUniqueInput`;
  }

  @Memoize()
  public get type(): GraphQLInputObjectType {
    assert(this.public, `"${this.name}" is private`);

    const componentSet = new Set<TComponent>(
      (<TComponent[]>[]).concat(
        ...[...this.node.publicUniqueConstraintSet].map(({ componentSet }) => [
          ...componentSet,
        ]),
      ),
    );

    return new GraphQLInputObjectType({
      name: this.name,
      description: `Identifies exactly one "${
        this.node.name
      }" node given a unique combination of values:\n${[
        ...this.node.publicUniqueConstraintSet,
      ]
        .map(({ componentSet }) =>
          [...componentSet].map(({ name }) => name).join(', '),
        )
        .join('\n')}`,
      fields: () =>
        Object.fromEntries(
          [...componentSet].map((component) => [
            component.name,
            {
              type: GraphQLNonNullDecorator(
                component instanceof Leaf
                  ? component.type
                  : component.to.whereUniqueInput.type,
                [
                  ...this.node.publicUniqueConstraintSet,
                ].every((uniqueConstraint) =>
                  uniqueConstraint.componentSet.has(component),
                ),
              ),
            },
          ]),
        ),
    });
  }

  protected parseComponentValue(
    component: TComponent,
    value: unknown,
    path: Path,
  ): ValueOf<TWhereUniqueInputValue> {
    if (value === undefined) {
      throw new UnexpectedUndefinedValueError(undefined, path);
    } else if (value === null) {
      if (!component.nullable) {
        throw new UnexpectedNullValueError(undefined, path);
      }

      return null;
    }

    return component instanceof Leaf
      ? assertLeafValue(component.type, path)
      : component.to.whereUniqueInput.parseValue(
          value as any,
          path,
          component.preferredUniqueConstraintSet,
        );
  }

  public parseValue(
    value: TWhereUniqueInputValue,
    path: Path = addPath(undefined, this.name),
    uniqueConstraints: Iterable<UniqueConstraint> = this.node.uniqueConstraintMap.values(),
  ): TParsedWhereUniqueInputValue {
    if (!isPlainObject(value)) {
      throw new UnexpectedValueError(value, `an object`, path);
    }

    for (const uniqueConstraint of uniqueConstraints) {
      try {
        return Object.fromEntries(
          Array.from(uniqueConstraint.componentSet, (component) => [
            component.name,
            this.parseComponentValue(
              component,
              value[component.name],
              addPath(path, component.name),
            ),
          ]),
        );
      } catch (error) {
        if (error instanceof UnexpectedValueError) {
          // Let's try the next "unique constraint"
          continue;
        } else {
          throw error;
        }
      }
    }

    throw new UnexpectedValueError(
      value,
      `an unique combination of value`,
      path,
    );
  }
}
