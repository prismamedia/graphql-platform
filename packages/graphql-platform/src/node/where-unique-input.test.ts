import { GraphQLInputObjectType, printType } from 'graphql';
import { GraphQLPlatform } from '..';
import { MyGP, nodeNames, nodes } from '../__tests__/config';
import { TWhereUniqueInputValue, WhereUniqueInput } from './where-unique-input';

describe('WhereUniqueInput', () => {
  let gp: MyGP;

  beforeAll(() => {
    gp = new GraphQLPlatform({ nodes });
  });

  it.each(nodeNames)('has a(n) %sWhereUniqueInput', (nodeName) => {
    const node = gp.getNode(nodeName);
    const whereUniqueInput = node.whereUniqueInput;

    expect(whereUniqueInput).toBeInstanceOf(WhereUniqueInput);
    expect(whereUniqueInput.public).toEqual(node.public);

    if (whereUniqueInput.public) {
      expect(whereUniqueInput.type).toBeInstanceOf(GraphQLInputObjectType);
      expect(
        printType(whereUniqueInput.type, { commentDescriptions: true }),
      ).toMatchSnapshot(whereUniqueInput.name);
    } else {
      expect(() => whereUniqueInput.type).toThrowError(
        `"${nodeName}WhereUniqueInput" is private`,
      );
    }
  });

  it.each([
    [
      'Article',
      undefined,
      `At ArticleWhereUniqueInput - expects an object, got: undefined`,
    ],
    [
      'Article',
      null,
      `At ArticleWhereUniqueInput - expects an object, got: null`,
    ],
    [
      'Article',
      { id_is_null: true },
      `At ArticleWhereUniqueInput - expects an unique combination of value, got: {\"id_is_null\":true}`,
    ],
    [
      'Article',
      { id: null },
      `At ArticleWhereUniqueInput - expects an unique combination of value, got: {\"id\":null}`,
    ],
    [
      'Article',
      { id: 123 },
      `At ArticleWhereUniqueInput - expects an unique combination of value, got: {\"id\":123}`,
    ],
    [
      'Article',
      { category: { parent: { title: null } } },
      `At ArticleWhereUniqueInput - expects an unique combination of value, got: {\"category\":{\"parent\":{\"title\":null}}}`,
    ],
  ])(
    'throws an Error on %sWhereUniqueInput.parseValue(%p)',
    (nodeName, value, error) => {
      const node = gp.getNode(nodeName);
      const whereUniqueInput = node.whereUniqueInput;

      expect(
        // @ts-expect-error
        () => whereUniqueInput.parseValue(value),
      ).toThrowError(error);
    },
  );

  it.each<[string, TWhereUniqueInputValue, TWhereUniqueInputValue]>([
    ['Article', { _id: 123 }, { _id: 123 }],
    [
      'Article',
      { id: 'e22205cc-7d8e-4772-a46d-528a29fb81f2' },
      { id: 'e22205cc-7d8e-4772-a46d-528a29fb81f2' },
    ],
    [
      'Article',
      { id: '3874e4bd-3b2b-456e-a811-7dd40b979f42', extraField: 'extraValue' },
      { id: '3874e4bd-3b2b-456e-a811-7dd40b979f42' },
    ],
    ['Category', { parent: null, order: 123 }, { parent: null, order: 123 }],
    [
      'Category',
      { parent: { parent: null, slug: 'root' }, order: 456 },
      { parent: { parent: null, slug: 'root' }, order: 456 },
    ],
    [
      'Category',
      { parent: { id: '6684d029-0016-4615-b1e7-f7f0087dbf11' }, order: 789 },
      { parent: { id: '6684d029-0016-4615-b1e7-f7f0087dbf11' }, order: 789 },
    ],
    [
      'Category',
      {
        parent: {
          parent: null,
          _id: 999,
          slug: 'root',
          id: 'd038d5cf-815e-4f8d-8099-1def0bdec246',
        },
        order: 789,
      },
      {
        parent: {
          _id: 999,
        },
        order: 789,
      },
    ],
  ])('%sWhereInput.parseValue(%p) = %p', (nodeName, value, parsedValue) => {
    const node = gp.getNode(nodeName);
    const whereUniqueInput = node.whereUniqueInput;

    expect(whereUniqueInput.parseValue(value)).toEqual(parsedValue);
  });
});
