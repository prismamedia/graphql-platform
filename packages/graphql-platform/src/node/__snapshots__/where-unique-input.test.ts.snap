// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WhereUniqueInput has a(n) ArticleTagWhereUniqueInput: ArticleTagWhereUniqueInput 1`] = `
"# Identifies exactly one \\"ArticleTag\\" node given a unique combination of values:
# article, tag
# article, order
input ArticleTagWhereUniqueInput {
  article: ArticleWhereUniqueInput!
  tag: TagWhereUniqueInput
  order: NonNegativeInt
}"
`;

exports[`WhereUniqueInput has a(n) ArticleWhereUniqueInput: ArticleWhereUniqueInput 1`] = `
"# Identifies exactly one \\"Article\\" node given a unique combination of values:
# id
input ArticleWhereUniqueInput {
  id: UUID!
}"
`;

exports[`WhereUniqueInput has a(n) CategoryWhereUniqueInput: CategoryWhereUniqueInput 1`] = `
"# Identifies exactly one \\"Category\\" node given a unique combination of values:
# id
# parent, slug
# parent, order
input CategoryWhereUniqueInput {
  id: UUID
  parent: CategoryWhereUniqueInput
  slug: NonEmptyTrimmedString
  order: NonNegativeInt
}"
`;

exports[`WhereUniqueInput has a(n) HitWhereUniqueInput: HitWhereUniqueInput 1`] = `
"# Identifies exactly one \\"Hit\\" node given a unique combination of values:
# _id
input HitWhereUniqueInput {
  _id: NonNegativeInt!
}"
`;

exports[`WhereUniqueInput has a(n) TagWhereUniqueInput: TagWhereUniqueInput 1`] = `
"# Identifies exactly one \\"Tag\\" node given a unique combination of values:
# id
# slug
input TagWhereUniqueInput {
  id: UUID
  slug: NonEmptyTrimmedString
}"
`;

exports[`WhereUniqueInput has a(n) UserProfileWhereUniqueInput: UserProfileWhereUniqueInput 1`] = `
"# Identifies exactly one \\"UserProfile\\" node given a unique combination of values:
# user
input UserProfileWhereUniqueInput {
  user: UserWhereUniqueInput!
}"
`;

exports[`WhereUniqueInput has a(n) UserWhereUniqueInput: UserWhereUniqueInput 1`] = `
"# Identifies exactly one \\"User\\" node given a unique combination of values:
# id
# username
input UserWhereUniqueInput {
  id: UUID
  username: NonEmptyTrimmedString
}"
`;
